### PR TITLE
[proof] Remove duplication in the proof save path.

### PR DESCRIPTION
**Pull request by ejgallego**
Further refactorings in the handling of the proof save path, including better handling of mutually-defined constants.

### CODE REVIEW

File: vernac/lemmas.ml
Comment (SkySkimmer) [inline vernac/lemmas.ml]: This function name should mention `admitted` since it's not for real proofs.

Comment (ejgallego) [reply vernac/lemmas.ml]: Done, thanks!

**Review by herbelin — APPROVED** [review]
This PR factorizes the definition declaration part of `finish_proved` with `declare_definition` and the assumption declaration in `finish_admitted` and `save_remaining_recthms` with a new `declare_axiom` in `declareDef`.

I did not check all details, but it seems a relevant step to indeed simplify the paths used to declare CIC objects. I myself have factorizations to come in `comAssumption.ml` and I'd be interested in having this merged first.

Are there other ongoing works (or suggestions) about this part of the code to synchronize with (@skyskimmer, @ppedrot, @maximedenes, ...)?

PS: Since it is already from October, I would suggest to lift it to a full PR and to merge. Continuation of this work can happen later.

**Comment by ejgallego** [comment]
> PS: Since it is already from October, I would suggest to lift it to a full PR and to merge. Continuation of this work can happen later.

Indeed, thanks for the review @herbelin ; I was holding this on trying to solve a couple of questions I have wrt declare path, but I will do a pass and try to get it ready soon.

**Comment by ejgallego** [comment]
Ok, I think this is more or less ready; I have further cleanups but better to postpone to other PRs.

File: vernac/lemmas.ml
Comment (SkySkimmer) [inline vernac/lemmas.ml]: Is this really the right way to work? It seems suspicious.

File: vernac/lemmas.ml
Comment (SkySkimmer) [inline vernac/lemmas.ml]: Declaredef should handle that.

File: vernac/lemmas.ml
Comment (ejgallego) [reply vernac/lemmas.ml]: Umm, no idea how to solve that right now; frankly I'm not sure I understand all the details of the way recthms are handled; note that this is not a regression in itself, IIANM we did normalize twice before too.

I guess we should just revisit the way we handle recthms in full; this seems to need a large refactoring tho; we would likely have to move the knowledge of that down `proof`. Would be worth it on the other hand, got any ideas?

[This is the kind of stuff we could really benefit from some whiteboard discussion IMO]

File: vernac/lemmas.ml
Comment (ejgallego) [reply vernac/lemmas.ml]: I added a comment; indeed the problem again is that old code was not doing this for the rest of the constants, so if we move this `DeclareDef` now we will different semantics for mutual assumptions [how is that even handled tho]

I'm gonna push a commit with you suggestion and see what happens.

Comment (ejgallego) [reply vernac/lemmas.ml]: Not sure the case admit + recthms is possible tho?

**Comment by ejgallego** [comment]
Still one particular bad offender is `comAssumption`...

File: vernac/lemmas.ml
Comment (SkySkimmer) [reply vernac/lemmas.ml]: Normalize twice is not too bad, but prepare_def also does minimize which I don't trust so much here, did we already do that twice?

Comment (ejgallego) [reply vernac/lemmas.ml]: We did not; do you have any suggestion how to handle this?

I guess we should go ahead and refactor the mutual fixpoint code even more; I have an idea how to do that but it will be a bit invasive; basically I'll introduce a mutual_entry type that keeps the invariants, then it is projected.

Comment (ejgallego) [reply vernac/lemmas.ml]: Yup, writing this cleanup, I like how it looks, should be ready in a few days.

Comment (ejgallego) [reply vernac/lemmas.ml]: Cleanup done, have a look.

**Comment by ejgallego** [comment]
Ok I've reworked the way mutual constants are handled, it seems to work and IMHO should be better than the current way, in particular we don't access the previously-saved constant anymore [which was super-hackish]

There are further refactorings to perform, but IMO if the current code is OK I think we can merge this and defer them to further PRs; anyways if I'm able to do some more I'll push them here.

Running a bench now https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/872/

**Comment by ejgallego** [comment]
> Running a bench now https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/872/

Bench seems OK, this needs assignee, I guess it could interest any of @coq/vernac-maintainers or @SkySkimmer

**Comment by ejgallego** [comment]
By the way @maximedenes you may be interested in https://github.com/coq/coq/pull/11016/commits/17327d50df477670ba9fafc30789cfa62ac6a8e3

In general I think that it is safe to say that we will be able to consolidate all the fixpoint hacks in a single place in declare "some time soon"(*), so indeed we shall be able to decide what to do in this sense a bit better.

(*) realistically speaking we are talking about a few dozen refactoring PRs, writing the code should take a few days, merging it, maybe months?

File: vernac/declareDef.ml
Comment (SkySkimmer) [inline vernac/declareDef.ml]: This means once we have the opaque `Lemma := ` it will also suggest, which is wrong. I think deciding to suggest should stay in lemmas.

File: vernac/lemmas.ml
Comment (SkySkimmer) [inline vernac/lemmas.ml]: Why "mutind"?

File: vernac/lemmas.ml
Comment (SkySkimmer) [inline vernac/lemmas.ml]: Should we assert i=0 in the non Mutual cases?

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
I don't like the "suggest" change.

File: vernac/declareDef.ml
Comment (ejgallego) [reply vernac/declareDef.ml]: IMO the whole "suggest" business still needs some work; after thinking a bit tho I'd suggest you consider we keep it in `DeclareDef` for now, here are 3 reasons:
- in order to work it is accessing the low-level `proof_entry` struct, which we do make fully abstract in upcoming PRs, thus the logic has to be moved to declare
- having it back in lemmas means that other users of `DeclareDef` with opaque don't get suggestions, for example obligations / mutual definitions. IMO this is wrong, as this suggestion should be uniform. You may be right that this could be triggering a few strange situations, but IMO the correct thing to do is to put this in the common path and fix the callers so they can all provide the hints.
- `Lemma :=` is still not fully determined so I'd rather wait until it arrives to take decisions

So I still prefer to keep it at this level, specially due to point 2, but if you still want me to move it back no problem.

File: vernac/lemmas.ml
Comment (ejgallego) [reply vernac/lemmas.ml]: I meant "mutually inductive definition" but I agree that's a bad name, correcting.

File: vernac/lemmas.ml
Comment (ejgallego) [reply vernac/lemmas.ml]: Sure why not, in principle nothing limits this code to be used in mutual, but not recursive defs tho.

Comment (ejgallego) [reply vernac/lemmas.ml]: I've added it but unsure this will remain once all code is unified [in depending PR]

**Comment by ejgallego** [comment]
Updated wrt comments; thanks. Regarding the suggest change see my comment, if you are still unconvinced I'll move it back to lemmas.

File: vernac/lemmas.ml
Comment (ejgallego) [reply vernac/lemmas.ml]: This is actually incorrect for mutual coinductives [which are handled this way]

Comment (ejgallego) [reply vernac/lemmas.ml]: Indeed it is not clear we are doing the right thing for cofixpoints; CI passes tho.

Comment (ejgallego) [reply vernac/lemmas.ml]: Seems we are doing the right thing, AFAICT.

**Comment by ejgallego** [comment]
> Regarding the suggest change see my comment, if you are still unconvinced I'll move it back to lemmas.

@SkySkimmer I did some more advance in the PRs that depend on this one and I think we can keep it there; the goal is to unify the interactive and non-interactive declaration mode, so we will be able to distinguish `Lemma foo : T.` from `Lemma foo := def.` as the API in declare will take two different objects.

We could indeed delay the move to #11785  but IMO not worth it.

**Review by SkySkimmer — APPROVED** [review]
OK then

**Comment by SkySkimmer** [comment]
Is the perennial failure related?

**Comment by ejgallego** [comment]
> Is the perennial failure related?

I think it was due to rewriter, let's re-run.

**Comment by ejgallego** [comment]
Failure in fiat-crypto look real, I'll have a look, this used to work but maybe I messed up something.

**Comment by ejgallego** [comment]
Ok, problem fixed, was a stupid mistake in the last commit; I didn't notice it in my working tree as upcoming commits remove the special handling for the "0" case in mutdef so the bug was gone.

**Pull request merged**
